src/sys.c

```diff
@@ -635,6 +635,38 @@ JL_DLLEXPORT long jl_SC_CLK_TCK(void)
 #endif
 }
 
+#ifdef _OS_OPENBSD_
+// Helper for jl_pathname_for_handle()
+struct dlinfo_data {
+    void       *searched;
+    const char *result;
+};
+
+static int dlinfo_helper(struct dl_phdr_info *info, size_t size, void *vdata)
+{
+    struct dlinfo_data *data = (struct dlinfo_data *)vdata;
+    void *handle;
+
+    /* ensure dl_phdr_info at compile-time to be compatible with the one at runtime */
+    if (sizeof(*info) < size)
+        return -1;
+
+    /* dlopen the name */
+    handle = dlopen(info->dlpi_name, RTLD_LAZY | RTLD_NOLOAD);
```

Comment:
This is a pretty big hammer for this implementation. Is there really no information in the dl_phdr_info that would let you make this conclusion?

Reply:
OpenBSD doesn't have `dlinfo()` function, so alternative implementation was necessary. The idea for the implementation comes from `libwhich`.

Here, we are walking across all the already loaded libraries (using `dl_iterate_phdr()`). We get the name of a library, and call `dlopen()`. The library is expected to be already loaded and `dlopen()` will just return the handler for it without too much processing. Next we check if the returned handle is the one we are searching, and if yes, the `struct dlinfo_data` data is accurate and we can extract the library name.

So in the general case, it isn't too bad (outside the O(n) search). Race might be still possible, and the worst case would be `dlopen()` will go on the filesystem to search and re-open the library just closed.